.github/workflows/ci.yaml

```diff
@@ -1,92 +1,171 @@
+      fail-fast: false
+      matrix:
+        # we build with a specific JDK version but source/target compatibility should ensure the jar is usable by
+        # the target versions we want to support
+        java-version:
+          - 8
+          - 11
+          - 17
+          - 21
     steps:
       - name: Checkout sources
         uses: actions/checkout@v4
 
-      - uses: actions/cache@v4
-        with:
-          path: |
-            ~/.gradle/caches
-            ~/.gradle/wrapper
-          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*') }}
-          restore-keys: |
-            ${{ runner.os }}-gradle-
-
       - name: Configure JDK
         uses: actions/setup-java@v3
         with:
           distribution: 'corretto'
           java-version: 17
           cache: 'gradle'
 
-      - name: Build and Test
+      - name: Setup build environment
+        uses: ./.github/actions/setup-build
+      - name: Test with ${{ matrix.java-version }}
+        shell: bash
         run: |
-          echo "kotlinWarningsAsErrors=true" >> $GITHUB_WORKSPACE/local.properties
-          ./gradlew allTests
+          ./gradlew jvmApiCheck
+          ./gradlew -Ptest.java.version=${{ matrix.java-version }} jvmTest --stacktrace
+      - name: Save Test Reports
+        if: failure()
+        uses: actions/upload-artifact@v4
+        with:
+          name: test-reports-jvm-${{ matrix.java-version }}
+          path: '**/build/reports'
 
-  macos-compat:
-    runs-on: macos-latest
+  # macos-14 build and test for targets: jvm, macosArm64, iosSimulatorArm64, watchosSimulatorArm65, tvosSimulatorArm64
+  # macos-13 build and test for targets: jvm, macosX64, iosX64, tvosX64, watchosX64
+  macos:
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [macos-14, macos-13]
+    runs-on: ${{ matrix.os }}
     steps:
-      - name: Checkout sources
-        uses: actions/checkout@v4
+      - uses: actions/checkout@v4
+        with:
+          submodules: true
+      - name: Setup build environment
+        uses: ./.github/actions/setup-build
+      - name: Build and Test ${{ env.PACKAGE_NAME }}
+        run: |
+          if [ "${{ matrix.os }}" = "macos-13" ]; then
+            ./gradlew build -PiosSimulatorDevice="iPhone 15"
+          else
+            ./gradlew build
+          fi
+      - name: Save Test Reports
+        if: failure()
+        uses: actions/upload-artifact@v4
+        with:
+          name: test-reports-macos-${{ matrix.os }}
+          path: '**/build/reports'
 
-      - uses: actions/cache@v4
+  # build and test for targets: jvm, linuxX64
+  # cross compile for: linuxX64, linuxArm64
+  linux:
+    runs-on: ubuntu-22.04
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: true
+      - name: Setup build environment
+        uses: ./.github/actions/setup-build
+      - name: Configure Docker Images
+        run: |
+          ./docker-images/build-images.sh linux-x64 linux-arm64
```

Comment:
note: @ianbotsf We are still using Docker for Linux CI, I will need to think about how we remove Docker later.